### ISSUE TITLE
AccessControl filter does not support RBAC rules

### ISSUE THREAD

**Issue by mrbig** [yiisoft/yii2#13598]
### Problem description

[RBAC rules](http://www.yiiframework.com/doc-2.0/guide-security-authorization.html#using-rules) are very useful to solve complex authorization problems. Unfortunately the AccessControl filter does not support them out of the box because custom parameters can not be fed to the roles. So one must use each time the matchCallback with custom implementation, but it results in such non elegant access rules where the real role check is hidden in the callback:

``` 
[
    'allow' => true,
    'roles' => ['@'],
    'matchCallback' => function($rule, $action) {
        return Yii::$app->user->can('updateOwnPost', ['postId' => 2]);
    }
]
```

### Proposed solution

Here I'm proposing a solution to add an optional ruleParamsCallback which should return any parameters necessary for the User::can check. For example:

```
[
    'allow' => true,
    'roles' => ['updateOwnPost'],
    'ruleParamsCallback' => function($rule, $action, $role) {
        return ['postId' => 12];
    }
]
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.12
| PHP version      | any
| Operating system | any

**Comment by mrbig** [comment]
Created pull request #13599 to demonstrate proposal

**Comment by arogachev** [comment]
Several duplicate issues already exist: #7643, #13273, #13552.

I think I explained well why this should not be implemented in https://github.com/yiisoft/yii2/issues/13273#issuecomment-269185026.

**Comment by dynasource** [comment]
I'm positive against improving the architecture for a better cooperation between RBAC & Accessfilter. Both are very related and this relation deserves a delicate solution in the framework. 

This PR https://github.com/yiisoft/yii2/pull/13599 covers one of the usecases, although its still some kind "monkey patching". It must be further developed into a general  sophisticated architecture.

**Comment by arogachev** [comment]
@dynasource How you see the handling of this problem then:

> Usually you need to find the model and throw according **404 Not Found** expection if it was not found first and only then throw **403 Forbidden** expection in case of lack of needed permissions. Where we need to write **404 Not Found** check in this approach? Putting this inside a RBAC rule is definetely incorrect, it's a controller's action job no doubt. With additional separate callback you lose integrity of the action and it's not absolutely clear what it does.

**Comment by dynasource** [comment]
good question. It's a problem most of us will recognize. The current state of the framework does not solve this issue out of the box. 

We should look for a solution in which RBAC and the Controller can communicate. Im thinking of a third Architecture which processes the $model for interventions such as exceptions like 403 / 404.

**Comment by arogachev** [comment]
This is not the only one problem that needs to be solved. I also mentioned example of more complicated checks.

I think having 3 lines in controller for checking permissions is perfectly fine and clear. Sometimes we should not try to solve everything through configuration.

**Comment by dynasource** [comment]
> I think having 3 lines in controller for checking permissions is perfectly fine and clear. Sometimes we should not try to solve everything through configuration.

The way I look at it:
* controller logic should be minimized, ideally be standardized and should only require customization when use cases are rare
* RBAC is common enough to require a default integration with controllers

The reason for looking at this subject from this perspective is to be able to efficiently manage an infinite number Modules with flexibility in mind. RBAC can be customized without having to alter Controllers implementations.

**Comment by arogachev** [comment]
This is exactly the kind of logic that should be inside a controller IMO. `AccessControl` is a handy feature no doubt, but if we need to pass the parameter to RBAC rule, and this parameter (model) is created and calculated inside a controller, why we need to move it anywhere else? MVC principle is not violated here, controller is just checking if data is allowed, if no - returns 403 Forbidden page, else - renders data in desired format. Also the length of such code is minimal.

The solution that might solve this problem - is something like a [Class Based Views] in Django. This allows to configure action (view in Django's terms) through overriding corresponding attributes and methods of a class. Applicable to Yii, maybe involve standalone actions? For example, we can have something like `UpdateOwnEntryAction`, pass the `ActiveQuery` instance and RBAC rule to it and it will handle these extra checks for us. Hovewer, as I said, things can be more complicated. I think this should not be included in the core neither.

The thing is, as a developer, just by a quick glance at the action, I want to understand what it does and what's inside of it. With logic moved out from action it will be much harder.

ACF is good for simple permissions, it's really useful when we need to check the same permission for multiple actions, etc.

But this: 

```php
if (!Yii::$app->user->can('entry.view', ['model' => $model])) {
    throw new ForbiddenHttpException("Sorry, you don't have permission to view this entry.");
}
```

is a specific action's thing. Why we need to isolate it somewhere in behavior?

[Class Based Views]: https://docs.djangoproject.com/en/1.10/topics/class-based-views/

**Comment by dynasource** [comment]
@arogachev, thanks for sharing your view.

It depends on which hat you put on. In my experience, the inner workings of Actions have proven to be so generic that these are easily standardizable. Looking at it is not necessary anymore for me to understand them because of standardization. 

> Why we need to isolate it somewhere in behavior?
Coming back to your question. Isolation is useful for:
* reuse (isolation enables reuse)
* simplicity (all actions are standardized)
* flexibility (central configuration lets you customize RBAC for i.e. modules)

**Comment by samdark** [comment]
One problem still remains. Before action such as `actionView($slug)` is executed you have no idea which model to find, how to do it and what to do in case it's not found. It depends very much on the application: in case of missing model you may want to suggest another model based on soundex distance of a slug or just display 404 or redirect back to homepage. Finding a model could be non-trivial as well if you store it in a special storage or something.

I saw one attempt standardizing such actions and checks. It was CodeIgniter's automagic CRUD which was then deprecated because it was fitting the case till the first deviation which happened all the time.

**Comment by dynasource** [comment]
> Problem still remains. Before action such as actionView($slug) is executed you have no idea which model to find, how to do it and what to do in case it's not found.

Im talking about a intervention class `after` the find has been completed. This class handles the `$model` result, which is either found or empty. Its up to the developer how this intervention class is configured. Whether the model should be passed to the RBAC component, an Error component, or perhabs an advanced Statistics component, but in most cases just continuation towards rendering the view.

**Comment by mrbig** [comment]
Hello guys,

I think you are overthinking it. In my point of view access control should be something that's stand alone, safe, and robust. IMHO ACF and RBAC are much closely related than RBAC and Actions. 

My customers want to oversee permissions around the application, audit and reconfigure them at any time.

Currently I have to create an 'open to all' ACF, then do the fine grained check in the action, and make sure that  every execution path is covered with access control requested by the costumer. @arogachev wrote that it's only three lines, but in practice it gets more and more complicated, and my issue is that there's no way to enforce it.

I'm also against sending query parameters directly to RBAC - that's not something secure. In my proposal you can use a closure to get the parameters. You could use anything: feed in the parameters directly, or look up a model, do something with it, or throw 404. Or you could create a custom implementation somewhere, that suits your configuration needs in more general.

There are several use cases (and mines also one of them), where rule information is inherent in the permission model. Most of the time we don't need to look up a model to validate an access rule.

I'm also positive to replace the closure with an interface that makes reuse more simple.

**Comment by arogachev** [comment]
> I think you are overthinking it. In my point of view access control should be something that's stand alone, safe, and robust.

I agree with it for simple permissions. For complex permissions that depends on other objects, I don't think so. Plus based on experience with other web frameworks, it's a normal practice. Maybe vice-versa, you are overthinking it?

> @arogachev wrote that it's only three lines, but in practice it gets more and more complicated

Could you provide an example of such complex check? Most likely you are just trying to put extra logic of checking permissions to controller? Just isolate it in RBAC rule and it will be almost always have the size of 3 lines.

If you are nevertheless want to try to optimize it further, as I said, take a look at standalone actions. Applicable to Yii, it might be a good solution. Create separate action something like `UpdateOwnEntryAction`, add the `ActiveQuery`, template to render and other parameters inside of it, do everything you need within this action and then include and configure it in `actions()` section of a controller. 

Still there will be some problems with code that needs to be additionally executed for the specific action, because you can't make such things generic for the whole application. But it can be solved with isolation of all related business logic to according services and calling them within callbacks and passing these callbacks as parameters to standalone action. Personally I don't like this way of writing actions for the whole application (unless they are truly generic), but maybe you just need to give it a try. As an example, you can take a look to @klimov-paul's extension https://github.com/yii2tech/admin and [standalone actions](https://github.com/yii2tech/admin/tree/master/actions) implemented there.

**Comment by mrbig** [comment]
Hello @argochev I didn't want to offense you. You admit as well, that ACF can not handle all the features that RBAC is capable of. Per your other comments I'm not the first one requesting to patch this. I prefer access control in "configuration" but we're not all the same. You could take this issue nothing more than a +1 if you like.

Regarding you proposal about standalone actions: I'm aware of them, and use them where appropriate, but share the same concerns in the case you also described. Before submitting this FR I've already solved my problem by overriding ACF and AccessRule classes - to my taste that was more elegant and more reusable.

@dynasource requested to rethink the problem in a more OO way. What if this code:
```
    public function allows($action, $user, $request)
    {
        if ($this->matchAction($action)
            && $this->matchRole($user)
            && $this->matchIP($request->getUserIP())
            && $this->matchVerb($request->getMethod())
            && $this->matchController($action->controller)
            && $this->matchCustom($action)
        ) {
            return $this->allow ? true : false;
        } else {
            return null;
        }
    }
```
gets replaced by a QueryInterface like pattern, where each match is a different class, and a rule is just a collection of these match patterns. Then adding custom matches or replacing one like the matchRole could be pretty straightforward.

That would open up several possibilities like like authentication to external resources, or implementing a rate limit all combined with already existing implementations (IP, Verb, Controller, Action check)

**Comment by dynasource** [comment]
> Hello @argochev I didn't want to offense you.
I tried to spot any offense but I couldn't find any ;) so no worries.

> gets replaced by a QueryInterface like pattern, where each match is a different class, and a rule is just a collection of these match patterns. Then adding custom matches or replacing one like the matchRole could be pretty straightforward.

sounds good!

**Comment by arogachev** [comment]
I'm not offended in any way, just trying to argue and explain my point of view.

From my personal experience, the most complicated permissions checks is when you need to allow action to certain object:

- for those who has super privileges for this type of object;
- for those who has privileges for this specific object.

The problem is solved easily with 2 dedicated permissions and check:

```php
if (!Yii::$app->user->can('entry.view.all') && !Yii::$app->user->can('entry.view', ['model' => $model])) {
    throw new ForbiddenHttpException("Sorry, you don't have permission to view this entry.");
}
```

All logic for `entry.view` is hidden inside according RBAC rule which can look for example like this:

```php
<?php

namespace frontend\rules\entries;

use yii\rbac\Rule;

class ViewRule extends Rule
{
    public $name = 'entry.view';

    public function execute($user, $item, $params)
    {
        /* @var $model \common\models\Entry */
        $model = $params['model'];

        return $model->is_active && !$model->isExpired() && $model->author_id == $user->id;
    }
}
```

If you feel that the first line is too long, you can refactor it to:

```php
$canViewAll = Yii::$app->user->can('entry.view.all');
if (!$canViewAll && !Yii::$app->user->can('entry.view', ['model' => $model])) {
    throw new ForbiddenHttpException("Sorry, you don't have permission to view this entry.");
}
```

Moreover, technically you can hide "all" check inside of more specific check (RBAC rule for `entry.view`):

```php
if ($user->can('entry.view.all')) {
    return true;
}
```

But the first one is more clear IMO.

You can go further with creating wrapper / shortcut, but there is nothing redundant in this code:

- permission names, they will vary depending on controller and action;
- `$model` is calculated above;
- throwing 403 Forbidden is perfecrtly fine, message needs to be customized depending on action too.

With configuration you need to write the same details - permission names, exception message and additionally handle finding of `$model`. What is the benefit to have it decoupled in other place and probably in more unclear and complicated way?

So what else you are trying to optimize here? Please give the real example of action with optimization.

Also, how this:

> My customers want to oversee permissions around the application, audit and reconfigure them at any time.

is related with this topic? With either approach you can use 3rd party extensons (there are tons of them) for this purpose or implement your own GUI for managing and assigning existing permissions.

**Comment by mrbig** [comment]
Yet again we're heading away from the original issue: how RBAC and ACF are related. Following your thoughts rule checks in ACF would be are absolutely unnecessary, since all of that could be done in the action level. In our team we prefer ACF over 'inline access check', because it's more separate, 'configuration style'.

Let me give you a bit more complex RBAC example, but before I should state, that I really like the RBAC model in Yii, so this issue is not about changing the RBAC.

Let's say you're having a blog engine, where users can register and run they own blogs. Then you have several blogs where each of them has an owner. Each blog has several posts, and those post can have several comments. Blogs, posts, and comment can be deleted by the blog owner wich you get by:

- $blog->ownerId
- $post->blog->ownerId
- $comment->post->blog->ownerId

As you can see there are different paths for blogs, posts and comments.

This leaves you a decision:

1. you create a single rule, where the parameter is the "blog owner". In this case you have to look up the owner before the access check is called
2. or you could create several rules where the parameter is the model, and each rule is different in how the owner is found from the current model
3. you automaticaly create rbac rules when a blog/post/comment is created, and put users into these groups

Each of the above have its pro's and cons. In the simplified blog example I would prefer #2, since we don't have so many actions. But in my current projects we're talking about hundreds of actions where most of them is related to a few entities, but the lookup path is different everywhere. Creating a separate class for each of them is a hassle, and makes code review more complicated..

The #3 approach I would avoid generally, because it makes very complicated to reorganize permissions later. For example: you want to allow comment deletion to the author of the comment => you should generate new items in the rbac graph.

My proposed patch is handy if you choose the #1 approach. In this case you could put the lookup method (Comment::find()->post->blog->ownerId) in the callback method, and if any of the items are not found you can issue a 404 or just return null/empty, which will cause a forbidden. You write only a single rbac rule and check there whether the logged in user is the blog owner.

Note: this is still an oversimplified example. In real life you would have guest owners, admins, local admins who are responsible for a given country etc...

**Comment by samdark** [comment]
@mrbig wouldn't that cause lookup even for actions where blog owner is not important?

**Comment by arogachev** [comment]
> In our team we prefer ACF over 'inline access check', because it's more separate, 'configuration style'.

This does not mean that it should be implemented at the core level. Again, using ACF is completely fine when you don't need any parameters which must be calculated in the action itself. ACF runs at the behavior level, before action is executed.

In your case you can just create the `AuthorRule` in RBAC, this example is even given in the [docs](http://www.yiiframework.com/doc-2.0/guide-security-authorization.html#using-rules) because it's pretty common. Then create the according permissions:

- `blog.delete`;
- `post.delete`;
- `comment.delete`.

and associate them with `AuthorRule` because the logic of accessibility of the model is common for all 3 types of objects and based on authorship. Optionally you can have super permissions for deleting all objects.

Then in action you write only:

```php
if (!Yii::$app->user->can('blog.delete', ['model' => $model])) {
    throw new ForbiddenHttpException("Sorry, you don't have permission to delete this blog.");
}
```

And that's it. I don't understand the hassle behind this you are talking about.

As an alternative technically you can use just one permission at module level (`blog`) for each blog item (blog, post, comment, etc.):

- `blog-item.delete`

if you feel that you don't need to divide permissions further.

**Comment by samdark** [comment]
I tend to agree with @arogachev. The code you're going to write into `ruleParamsCallback` is the same as the code to be put into `AuthorRule`. The difference is that `ruleParamsCallback` would clutter behaviors section of controller and be executed not only on actions actually ran but at every action hurting performance.

**Comment by mrbig** [comment]
@samdark no, if you split in separate ACF rules where parameters are needed and those where don't. Mostly you need to check different roles, so you can't put them into a single ACF rule anyway.

@arogachev 
> I don't understand the hassle behind this you are talking about

Let's look at the comment deletion. The author of the comment is not the same who is the blog owner. So if you meant to write an AuthorRule like the one you wrote in comment #281135672 then you won't be able to use it here, but a slightly altered version like this:

```
    public function execute($user, $item, $params)
    {
        /* @var $model \common\models\Entry */
        $model = $params['model'];

        $owner = $model->post->blog; // <== lookup happens here

        // This part is important and shoud be written once:
        return $owner->is_active && !$owner->isExpired() && $owner->author_id == $user->id;
    }
```

So the owner lookup is different, and it could be different for many other entities as well. In real life you have blog co authors as well, so the post author is not necessarily the only person having delete rights, but other related entities could have that also. You'll have images, uploaded content, logs, etc in a complex entity scheme. This way you'll end up having several rules, where lots of code get duplicated if you're not careful enough to refactor everything that's common. So you're at the decision with my previous solutions #1, #2, #3. 

But my question remains: if ACF can be made better, why don't improve it?

**Comment by mrbig** [comment]
@samdark access check is done before role check. Take a look at [the source](https://github.com/yiisoft/yii2/blob/master/framework/filters/AccessRule.php#L108`)

**Comment by samdark** [comment]
@mrbig right. Then my performance point isn't valid.

**Comment by samdark** [comment]
So... for example, there's an action like the following. That's a common scenario. How would you implement that w/ access behavior?

```
public function actionView($id)
{
    $model = Post::find()->where(['id' => 10, 'status' => Post::STATUS_ACTIVE])->one();
    if (!$model) {
        // throw 404
    }
    
    if (!Yii::$app->user->can('view_post', ['post' => $post])) {
        // throw access denied if not permitted
    }

   // do the rest of the action
```

**Comment by samdark** [comment]
Note that `view_post` may be assigned either directly if you're an admin or via `view_own_post` which includes a rule if you're regular user.

**Comment by mrbig** [comment]
@samdark first let me assure you that I'm not arguing in favor for my current solution. I agree @dynasource that it can be implemented better a more OO way.

With the current implementation I'd do something like this:

```
'rules' => [
    [
        'actions' => ['view'],
        'allow' => true,
        'roles' => ['view_post'],
        'ruleParamsCallback' => function($rule, $role) {
            return [
                'post' => Post::find()->where(['id' => 10])->one();
            ]
        }
    ]
]
```

Remarks:

1. status check is done in the rule. This makes sure that it never will be forgotten
2. throwing a 403 is acceptable in this case if the post is not found. If not then the 404 should be thrown inside the callback, wich I admit is not the nicest way.
3. the closure can be factored away elsewhere if it needs to be reused, or for readability

What I like in a solution like this is that if you look at the ACF rule, you'll see the exact role that's validated, and not a general '@' role. IMO this is easier to read when auditing a controller with several actions.

**Comment by dynasource** [comment]
it all depends on how one wants to configure his application. You can either configure centrally or decentrally.

@mrbig and I (seem to) prefer the first. In order to achieve what we want, we currently can use Events. However, ideally there should be a 'application lifecycle` controller which makes such an approach more verbose.

**Comment by mrbig** [comment]
@dynasource I like in Yii that it doesn't put too much constraints on how your action lifecycle is built. The current behavior/filter based solution is simple and leaves the decision up to the user.

By rethinking I meant that the current AccessControl and AccessRule class could be reorganized in a way that different checks could be added by the user.

**Comment by samdark** [comment]
@mrbig so in the action you're going to fetch model again?

**Comment by mrbig** [comment]
@samdark in this example yes, but this is not an issue if you have proper in memory caching set up.

And there could be use cases when you don't need to fetch the model at all, because your permissions are stored somewhere else. I believe this was the tought behind @slinstj proposal in issue #13273 

IMHO you always have to balance between readability/managability and performance. Security is something where lot of us will sacrifice some performance, so I would leave this decision up to the user.

**Comment by samdark** [comment]
Well, to me a simple check in the action itself which doesn't require extra caching, additional callback, handles 404 well etc. etc. looks much simpler and easier.

**Comment by mrbig** [comment]
@samdark this is a design decision highly dependent on the current project and personal taste.

No problem if you close the pull request #13598 but would love to see some kind of support for RBAC rules in ACF in the upcoming yii versions. You can take this as a +1 vote

**Comment by dynasource** [comment]
> would love to see some kind of support for RBAC rules in ACF in the upcoming yii versions. You can take this as a +1 vote

Thanks for sharing your idea's. We will keep this one open and I will pick this on up in the future.
